control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -165,6 +166,24 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 		}
 	}
 
+	// Determine if active control plane operator image matches the expected release image payload. If it doesn't
+	// need to no-op until the operator is recreated and at the matching version
+	if _, ok := hostedControlPlane.Annotations[hyperv1.ActiveHypershiftOperatorImage]; !ok {
+		return ctrl.Result{}, fmt.Errorf("active hypershift operator image annotation not present")
+	}
+	pullSecret := common.PullSecret(hostedControlPlane.Namespace)
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(pullSecret), pullSecret); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to lookup pull secret: %w", err)
+	}
+	expectedImage, err := cpoutil.GetHypershiftComponentImage(ctx, hostedControlPlane.Annotations, hostedControlPlane.Spec.ReleaseImage, r.ReleaseProvider, hostedControlPlane.Annotations[hyperv1.ActiveHypershiftOperatorImage], pullSecret.Data[corev1.DockerConfigJsonKey])
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to lookup expected image: %w", err)
+	}
+	if expectedImage != r.ActiveImage {
+		r.Log.Info("No-oping until component is recreated with proper image", "expectedImage", expectedImage, "activeImage", r.ActiveImage)
```

Comment:
Is it safe to abruptly short-circuit like this? i.e wouldn't we need to check the HCP.Status as well to validate it's stable and no upgrades in-place so it's safe to short-circuit?

Reply:
It's safe and should occur:  We should not be updating resources until we are at the appropriate matching version. The one caveat is delete since at that point the hypershift operator will no longer reconcile the control plane operator. So the delete path should be allowed to continue as is.